### DISCUSSION TITLE
How to access to neighbor list in CustomAction

### DISCUSSION THREAD

**Discussion by kojikokojiko** [glotzerlab/hoomd-blue#1675] [Support] [answered]
I would like to add the properties of the Vicsek model to a system using the WCA potential in NVE. For this purpose, I want to perform the averaging of the direction of the velocities of neighboring particles within a Custom Action. To improve the efficiency of the calculation, I would like to reuse the NeighborList used in Hoomd. I would appreciate it if you could teach me how to do this.

**Comment by joaander** [comment, marked as answer]
I recommend implementing a [C++ component that extends HOOMD-blue](https://github.com/glotzerlab/hoomd-component-template) for the best performance.

In Python, you can use [`gpu_local_nlist_arrays`](https://hoomd-blue.readthedocs.io/en/v4.4.0/module-md-nlist.html#hoomd.md.nlist.NeighborList.gpu_local_nlist_arrays]), [`cpu_local_nlist_arrays`](https://hoomd-blue.readthedocs.io/en/v4.4.0/module-md-nlist.html#hoomd.md.nlist.NeighborList.cpu_local_nlist_arrays), or [`local_pair_list`](https://hoomd-blue.readthedocs.io/en/v4.4.0/module-md-nlist.html#hoomd.md.nlist.NeighborList.local_pair_list)

**Reply by kojikokojiko** [reply]
Thank you for your response. I have read the documentation and understand that I can access the neighborlist using gpu_local_nlist_arrays. However, I am not sure how to call gpu_local_nlist_arrays in the definition of my custom CustomAction class. Could you please explain this in more detail?

**Reply by joaander** [reply]
As shown in the example snippet in the documentation, you access the local arrays as a context manager:
```
with nlist.gpu_local_nlist_arrays as data:
    # your code here
```

**Reply by kojikokojiko** [reply]
Thank you for your reply. Based on your advice, I have created this code. With this code, am I correctly obtaining the Neighborlist for the specified step?

```
class Viscek(hoomd.custom.Action):
    def __init__(self,nlist):
        self.nlist=nlist
        
    def act (self,timestep):
  
        with self.nlist.cpu_local_nlist_arrays as arrays:
            nlist_iter = zip(arrays.head_list, arrays.n_neigh)
            # print(timestep,file=sys.stderr)
        
            for i, (head, pair_length) in enumerate(nlist_iter):
                for j_idx in range(head, head + pair_length):
                ＃　my code




cell = hoomd.md.nlist.Cell(buffer=0.4)
lj = hoomd.md.pair.LJ(nlist=cell, mode="shift")
lj.params[("Move", "Move")] = dict(epsilon=epsilon_aa, sigma=sigma_aa)
lj.r_cut[("Move", "Move")] = 2**(1/6)*sigma_aa

update_time=100
vicsek_operation=hoomd.update.CustomUpdater(action=Viscek(cell),trigger=update_time)
```
